dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/EventRequiredPropertiesValidationHook.java

```diff
@@ -0,0 +1,188 @@
+package org.hisp.dhis.tracker.validation.hooks;
+
+import com.google.common.base.Preconditions;
+import org.hisp.dhis.organisationunit.OrganisationUnit;
+import org.hisp.dhis.program.*;
+import org.hisp.dhis.trackedentity.TrackedEntityInstance;
+import org.hisp.dhis.tracker.bundle.TrackerBundle;
+import org.hisp.dhis.tracker.domain.Event;
+import org.hisp.dhis.tracker.preheat.PreheatHelper;
+import org.hisp.dhis.tracker.report.TrackerErrorCode;
+import org.hisp.dhis.tracker.report.TrackerErrorReport;
+import org.hisp.dhis.tracker.report.ValidationErrorReporter;
+import org.hisp.dhis.user.User;
+import org.springframework.stereotype.Component;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Objects;
+
+import static org.hisp.dhis.tracker.report.ValidationErrorReporter.newReport;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+@Component
+public class EventRequiredPropertiesValidationHook
+    extends AbstractTrackerValidationHook
+{
+
+    @Override
+    public int getOrder()
+    {
+        return 300;
+    }
+
+    @Override
+    public List<TrackerErrorReport> validate( TrackerBundle bundle )
+    {
+        ValidationErrorReporter reporter = new ValidationErrorReporter( bundle, this.getClass() );
+        User actingUser = bundle.getPreheat().getUser();
+
+        for ( Event event : bundle.getEvents() )
+        {
+            reporter.increment( event );
+
+            boolean exists = programStageInstanceService.programStageInstanceExistsIncludingDeleted( event.getEvent() );
```

Comment:
this is a potential performance killer .... what we did sometime back was to fetch events in batch.

Reply:
Switched to look via preheat